Storage limits for store and forward messages

neonknight64 opened the pull request:
## Description
- These changes allow the storage capacity of the store and forward system to be configured and enforced. When the storage limit is reached then the oldest messages are discarded to preserve the configured storage limit.

## Motivation and Context
These changes ensure that the number of store and forward messages do not grow too large.

## How Has This Been Tested?
The truncate_message test was added to test the enforcement of the message storage limit.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

sdbondi commented:
Looks good - the SAF actor periodically (default every 10 mins) does a clean up of messages according to priority. I don't think we should do counts and cleanup on every insert but add a cleanup on the periodic check. The database size can go over the limit for a few minutes as long as it cleans up periodically.

sdbondi commented on comms/dht/src/store_forward/database/mod.rs:
```diff
@@ -259,13 +263,30 @@ impl StoreAndForwardDatabase {
-                    .map_err(Into::into)
+                    .execute(conn)?;
+
+                let msg_count = stored_messages::table
+                    .select(dsl::count(stored_messages::id))
+                    .first::<i64>(conn)? as usize;
+                if msg_count > storage_capacity {
+                    let remove_count = msg_count - storage_capacity;
+                    let message_ids: Vec<i32> = stored_messages::table
+                        .select(stored_messages::id)
+                        .order_by(stored_messages::stored_at.asc())
+                        .limit(remove_count as i64)
+                        .get_results(conn)?;
+                    num_removed += diesel::delete(stored_messages::table)
+                        .filter(stored_messages::id.eq_any(message_ids))
+                        .execute(conn)?;
+                }
```
I would do this in a new function (something like `truncate_messages(max_size: usize)`) and call it from the saf service - there's no extra query cost and it's better for a function to do one thing

sdbondi replied:
Also can you not do ordering and limit with a delete in one query?

`DELETE FROM stored_messages ORDER BY stored_at ASC LIMIT {remove_count}`

neonknight64 replied:
I changed it to make use of a `truncate_messages` function. I also attempted to perform the ordering and limit delete in one query, but was unable to get it to work.

Commit:
- These changes ensure that the number of store and forward messages do not grow too large.
- These changes allow the storage capacity of the store and forward system to be configured and enforced. When the storage limit is reached then the oldest messages are discarded to preserve the configured storage limit.
- The truncate_message test was added to test the enforcement of the message storage limit.

sdbondi reviewed (approved):
Nice 👌
